README.md
Provides an interface to the Open Platform's Content API of the Guardian Media Group. It retrieves content from news outlets The Observer, The Guardian, and guardian.co.uk from 1999 to current day.

ChangeLog
2013-05-07 - version 0.1:
   - Release
2014-02-17 - version 0.2:
   - Fixed the error caused by empty replies from Guardian API
   - Excluded duplicate entries from get_guardian and get_guardian_full
   - Revised documentation to explain search for exact keywords
   - Added BibTeX citation file 
   - Packages RCurl and RJSONIO are now on "Imports" instead of "Depends"
   - Fixed truncated lines in the PDF manual
2014-02-20 - version 0.3:
   - Removed the API-key provided for function get_guardian_full. Users should request an API-key at http://guardian.mashery.com/
2014-06-13 - version 0.4:
   - Function get_guardian_full declared in NAMESPACE. Users still need to request an API-key at http://guardian.mashery.com/
   - Functions get_json and parse_json_to_df are also declared in NAMESPACE
2014-09-16 - version 0.5:
   - The wrapper now complies with Guardian Content API v2
   - Search without an API-key is no longer supported in the new API. The functions get_guardian_full and get_guardian were merged into get_guardian, which now returns the full text
   - Function get_guardian requires an API key due to functional requirements of Guardian Content API v2
   - Guardian Content API v2 is faster, but implements a limit of 12 Calls per second and 5,000 calls per day
2016-03-11 - version 0.6:
   - Completed migration from API 2.0 beta to content.guardianapis.com
   - Added warning for requests over extended periods of time. Users should consider splitting requests in weekly or monthly chunks
   - cacert.pem now downloads (when needed) using HTTPS
2016-09-16 - version 0.7:
   - Declared dependency R >= 3.2.0
   - Fixed bug that returned only the second page of a search
   - Added section argument to get_guardian to allow searching articles within sections
   - Changed page-size from 50 to 100 to speed up API response
2016-10-18 - version 0.8:
   - Fixed pagination to match changes introduced to Guardian API 
2018-05-30 - version 0.9:
   - Updated Guardian API key for testing

R/guardianr.R
### Guardian API Wrapper v.09
### by M.T. Bastos & C. Puschmann
### contributions by Mark Johnman
# 
# # install these packages first if needed
# library(RCurl)
# library(RJSONIO)

# query Guardian API
get_json <- function(keywords, section, format, from.date, to.date, api.key)
{
# pagination
page.size <- 100
this.page <- 1
pages <- 1

if(as.Date(as.character(to.date))-as.Date(as.character(from.date))>31) {
    warning("The requested period is too long. Make weekly or monthly requests to avoid API interruptions")
}

# prepare list for storing api responses
api.responses <- NULL

# call guardian API
while (this.page <= pages) {
    if(is.null(section)) {
        request <- paste("http://content.guardianapis.com/search?q=", keywords, "&from-date=", from.date, "&to-date=", to.date, 
                         "&format=", format, "&show-fields=all&page=", this.page, "&page-size=", page.size, "&api-key=", api.key, sep="")
    } else {
        request <- paste("http://content.guardianapis.com/search?q=", keywords, "&section=", section, "&from-date=", from.date, "&to-date=", to.date, 
                         "&format=", format, "&show-fields=all&page=", this.page, 
                     "&page-size=", page.size, "&api-key=", api.key, sep="") }
    # query api
    if(.Platform$OS.type == "windows") { if(!file.exists("cacert.perm")) download.file(url="https://curl.haxx.se/ca/cacert.pem", destfile="cacert.perm") }
    if(.Platform$OS.type == "windows") { json <- getURL(request, cainfo = "cacert.perm", timeout = 240, ssl.verifypeer = FALSE) }
    else { json <- getURL(request, timeout = 240) }
    json <- fromJSON(json, simplify=FALSE)
    this.api.response <- json$response
    stopifnot(!is.null(this.api.response))
    #if(this.page==1) { pages.total <<- this.api.response$pages }
    if(this.api.response$total==0){
        print(paste("No matches were found in the Guardian database for keyword '", keywords, "'", sep=""))
        this.page <- this.page + 1
    } else {
        stopifnot(!is.null(this.api.response))
        pages <- this.api.response$pages
        if (pages >= 1)
        {
            print(paste("Fetched page #", this.page, " of ", pages, sep=""))
            api.responses <- c(api.responses, this.api.response)
        }
        else
        {
            print("Fetched page #1 of 1.")
        }
        api.responses <- c(api.responses, this.api.response)
        this.page <- this.page + 1
    }
}
return(api.responses)
if(.Platform$OS.type == "windows") { file.remove("cacert.perm") }
}

# parse json to data frame
parse_json_to_df <- function(api.responses)
{
api.df = data.frame(
    id=NULL,
    sectionId=NULL,
    sectionName=NULL,
    webPublicationDate=NULL,
    webTitle=NULL,
    webUrl=NULL,
    apiUrl=NULL,
    newspaperPageNumber=NULL,
    trailText=NULL,
    headline=NULL,
    showInRelatedContent=NULL,
    lastModified=NULL,
    hasStoryPackage=NULL,
    score=NULL,
    standfirst=NULL,
    shortUrl=NULL,
    wordcount=NULL,
    commentable=NULL,
    allowUgc=NULL,
    isPremoderated=NULL,
    byline=NULL,
    publication=NULL,
    newspaperEditionDate=NULL,
    shouldHideAdverts=NULL,
    liveBloggingNow=NULL,
    commentCloseDate=NULL,
    body=NULL)

# determine number of pages
pages <- which(names(api.responses)=="results")
#pages <- pages.total

for (i in pages)
{
    for (j in 1:length(api.responses[i]$results))
    {
        id <- api.responses[i]$results[[j]]$id
        sectionId <- api.responses[i]$results[[j]]$sectionId
        sectionName <- api.responses[i]$results[[j]]$sectionName
        webPublicationDate <- api.responses[i]$results[[j]]$webPublicationDate
        webTitle <- api.responses[i]$results[[j]]$webTitle
        webUrl <- api.responses[i]$results[[j]]$webUrl
        apiUrl <- api.responses[i]$results[[j]]$apiUrl
        newspaperPageNumber <- api.responses[i]$results[[j]]$fields$newspaperPageNumber
        trailText <- api.responses[i]$results[[j]]$fields$trailText
        headline <- api.responses[i]$results[[j]]$fields$headline
        showInRelatedContent <- api.responses[i]$results[[j]]$fields$showInRelatedContent
        lastModified <- api.responses[i]$results[[j]]$fields$lastModified
        hasStoryPackage <- api.responses[i]$results[[j]]$fields$hasStoryPackage
        score <- api.responses[i]$results[[j]]$fields$score
        standfirst <- api.responses[i]$results[[j]]$fields$standfirst
        shortUrl <- api.responses[i]$results[[j]]$fields$shortUrl
        wordcount <- api.responses[i]$results[[j]]$fields$wordcount
        commentable <- api.responses[i]$results[[j]]$fields$commentable
        allowUgc <- api.responses[i]$results[[j]]$fields$allowUgc
        isPremoderated <- api.responses[i]$results[[j]]$fields$isPremoderated
        byline <- api.responses[i]$results[[j]]$fields$byline
        publication <- api.responses[i]$results[[j]]$fields$publication
        newspaperEditionDate <- api.responses[i]$results[[j]]$fields$newspaperEditionDate
        shouldHideAdverts <- api.responses[i]$results[[j]]$fields$shouldHideAdverts
        liveBloggingNow <- api.responses[i]$results[[j]]$fields$liveBloggingNow
        commentCloseDate <- api.responses[i]$results[[j]]$fields$commentCloseDate
        body <- api.responses[i]$results[[j]]$fields$body
        
        if (is.null(id)) id <- NA
        if (is.null(sectionId)) sectionId <- NA
        if (is.null(sectionName)) sectionName <- NA
        if (is.null(webPublicationDate)) webPublicationDate <- NA
        if (is.null(webTitle)) webTitle <- NA
        if (is.null(webUrl)) webUrl <- NA
        if (is.null(apiUrl)) apiUrl <- NA
        if (is.null(newspaperPageNumber)) newspaperPageNumber <- NA 
        if (is.null(trailText)) trailText <- NA
        if (is.null(headline)) headline <- NA
        if (is.null(showInRelatedContent)) showInRelatedContent <- NA 
        if (is.null(lastModified)) lastModified <- NA
        if (is.null(hasStoryPackage)) hasStoryPackage <- NA
        if (is.null(score)) score <- NA
        if (is.null(standfirst)) standfirst <- NA
        if (is.null(shortUrl)) shortUrl <- NA
        if (is.null(wordcount)) wordcount <- NA
        if (is.null(commentable)) commentable <- NA
        if (is.null(allowUgc)) allowUgc <- NA
        if (is.null(isPremoderated)) isPremoderated <- NA
        if (is.null(byline)) byline <- NA
        if (is.null(publication)) publication <- NA
        if (is.null(newspaperEditionDate)) newspaperEditionDate <- NA
        if (is.null(shouldHideAdverts)) shouldHideAdverts <- NA
        if (is.null(liveBloggingNow)) liveBloggingNow <- NA
        if (is.null(commentCloseDate)) commentCloseDate <- NA
        if (is.null(body)) body <- NA
        
        this.api.df <- data.frame(
            id=id,
            sectionId=sectionId,
            sectionName=sectionName,
            webPublicationDate=webPublicationDate,
            webTitle=webTitle,
            webUrl=webUrl,
            apiUrl=apiUrl,
            newspaperPageNumber=newspaperPageNumber,
            trailText=trailText,
            headline=headline,
            showInRelatedContent=showInRelatedContent,
            lastModified=lastModified,
            hasStoryPackage=hasStoryPackage,
            score=score,
            standfirst=standfirst,
            shortUrl=shortUrl,
            wordcount=wordcount,
            commentable=commentable,
            allowUgc=allowUgc,
            isPremoderated=isPremoderated,
            byline=byline,
            publication=publication,
            newspaperEditionDate=newspaperEditionDate,
            shouldHideAdverts=shouldHideAdverts,
            liveBloggingNow=liveBloggingNow,
            commentCloseDate=commentCloseDate,
            body=body)		
        
        api.df <- rbind(api.df, this.api.df)
    }
}
return(unique(api.df))
}

# call api wrapper
get_guardian <- function(keywords, section=NULL, format="json", from.date, to.date, api.key)
{
guardian.api.responses <- get_json(keywords, section, format, from.date, to.date, api.key)
guardian.api.df <- parse_json_to_df(guardian.api.responses)
return (guardian.api.df)
}

MTB_-_GuardianR_Tutorial.R
## GuardianR Tutorial ##
## MT Bastos 14/05/13 ##

# set working folder
# setwd("")

# install and load package
# install.packages("GuardianR")
library(GuardianR)

# check help files
?get_guardian

# search guardian
results <- get_guardian(keywords="thatcher", from.date="2013-04-09", to.date="2013-04-09")

# search guardian for three months 
results <- get_guardian(keywords="thatcher", from.date="2012-01-09", to.date="2013-04-09")

# try different keywords
results <- get_guardian(keywords="World+Cup+Brazil", from.date="2013-01-09", to.date="2013-04-09")

# search guardian for three months
results <- get_guardian(keywords="World+Cup+Brazil", from.date="2012-01-09", to.date="2013-04-09")

# retrieve full text of guardian articles
# Guardian API-key can be obtained by registering at http://guardian.mashery.com/
results <- get_guardian_full(keywords="World+Cup+Brazil", from.date="2012-01-09", to.date="2013-04-09", api.key=api.key)

# check full article of entry #7
results$body[7]

# search for corinthians
results <- get_guardian_full(keywords="corinthians+brazil", from.date="2012-01-09", to.date="2013-04-09", api.key=api.key)

# search for corinthians again (no full text)
results <- get_guardian(keywords="corinthians+brazil", from.date="2010-01-09", to.date="2013-04-09")

# read headlines
results[,10]

# DATA ANALYSIS
# install.packages("lubridate")
library(lubridate)

# convert ISO8601 format to POSIX
results$webPublicationDate <- ymd_hms(results$webPublicationDate, quiet = FALSE, tz = "UTC", locale = Sys.getlocale("LC_TIME"), truncated = 0)
results$lastModified <- ymd_hms(results$lastModified, quiet = FALSE, tz = "UTC", locale = Sys.getlocale("LC_TIME"), truncated = 0)
results$newspaperEditionDate <- ymd_hms(results$newspaperEditionDate, quiet = FALSE, tz = "UTC", locale = Sys.getlocale("LC_TIME"), truncated = 0)
results$commentCloseDate <- ymd_hms(results$commentCloseDate, quiet = FALSE, tz = "UTC", locale = Sys.getlocale("LC_TIME"), truncated = 0)

# save results in rda and csv
# save(results,file="guardian_data.rda")
# write.csv2(results, file="guardian_data.csv")

# exploratory plot - publication date
par(mfrow=c(1,2))
plot(results$webPublicationDate, as.numeric(as.character(results$wordcount)), col="blue",cex=1.5,xlab="Web Publication Time", ylab="Article Word Count")
plot(results$webPublicationDate, as.numeric(as.character(results$score)), col="blue",cex=1.5,xlab="Web Publication Time", ylab="Article Score")

# exploratory plot - newspaper sections
par(mfrow=c(1,2))
plot(results$sectionName, as.numeric(as.character(results$wordcount)), col="lightblue",cex=2.5,xlab="Newspaper Sections", ylab="Article Word Count")
barplot(as.numeric(as.character(results$wordcount, results$sectionName)),  horiz=TRUE, names.arg=results$sectionName)

# exploratory plot - comparison between sister papers
par(mfrow=c(1,2))
plot(as.numeric(as.character(results$wordcount))~results$publication, col="blue",xlab="Sister Newspaper", ylab="Article Word Count", main="Word Count per Sister Newspaper")
plot(table(results$publication), main="Number of Articles per Sister Newspaper", ylab="")

# exploratory plot - time series
# install.packages("ggplot2")
library(ggplot2)
results$wordcount <- as.numeric(as.character(results$wordcount))
ggplot(results)+geom_point(aes(x=webPublicationDate,y=wordcount)) + labs(x = "time", y = "wordcount", title="Time Series") + theme(axis.text.x = element_text(colour="grey20",size=18,angle=0,hjust=1,vjust=1))

# install.packages("wordcloud")
# install.packages("tm")
library(wordcloud)
library(tm)

# plot wordcloud
if(NROW(results)>5000) {
  results.sp<-sample(results, size=5000)
  results.corpus <- Corpus(VectorSource(results.sp$headline))
} else {
  results.corpus <- Corpus(VectorSource(results$headline))
}
results.corpus <- tm_map(results.corpus, tolower)
results.corpus <- tm_map(results.corpus, removePunctuation)
results.corpus <- tm_map(results.corpus, removeWords, stopwords("english"))
print(wordcloud(results.corpus,colors=brewer.pal(6,"Dark2"),random.order=FALSE))

# plot wordcloud of full text
results <- get_guardian_full(keywords="corinthians+brazil", from.date="2012-01-09", to.date="2013-04-09", api.key=api.key)
if(NROW(results)>5000) {
  results.sp<-sample(results, size=5000)
  results.corpus <- Corpus(VectorSource(results.sp$body))
} else {
  results.corpus <- Corpus(VectorSource(results$body))
}
results.corpus <- tm_map(results.corpus, tolower)
results.corpus <- tm_map(results.corpus, removePunctuation)
results.corpus <- tm_map(results.corpus, removeWords, stopwords("english"))
print(wordcloud(results.corpus,colors=brewer.pal(6,"Dark2"),random.order=FALSE))

# check results with maximum time frame
results <- get_guardian(keywords="corinthians+brazil", from.date="2000-01-09", to.date="2013-04-09")
results$webPublicationDate <- ymd_hms(results$webPublicationDate, quiet = FALSE, tz = "UTC", locale = Sys.getlocale("LC_TIME"), truncated = 0)
results$wordcount <- as.numeric(as.character(results$wordcount))
ggplot(results)+geom_point(aes(x=webPublicationDate,y=wordcount)) + labs(x = "time", y = "wordcount", title="Time Series") + theme(axis.text.x = element_text(colour="grey20",size=18,angle=0,hjust=1,vjust=1))
plot(results$webPublicationDate, results$wordcount)

# check results with maximum time frame
results <- get_guardian(keywords="sao+paulo+football", from.date="2000-01-09", to.date="2013-04-09")
results$webPublicationDate <- ymd_hms(results$webPublicationDate, quiet = FALSE, tz = "UTC", locale = Sys.getlocale("LC_TIME"), truncated = 0)
results$wordcount <- as.numeric(as.character(results$wordcount))
ggplot(results)+geom_point(aes(x=webPublicationDate,y=wordcount)) + labs(x = "time", y = "wordcount", title="Time Series") + theme(axis.text.x = element_text(colour="grey20",size=18,angle=0,hjust=1,vjust=1))
plot(results$webPublicationDate, results$wordcount)

# check summary of results
summary(results)

# package citaton
citation("GuardianR")
